Increase safety of marshaled function closures (fixes issue #5942)

xavierleroy opened the pull request:
When a function value is marshaled (serialized) using the functions from module `Marshal`, the code of the function closure is not included in the marshaled data.  Instead, the code pointer contained in the closure is represented by a pair of
- a MD5 digest of a code block (e.g. the code for the main executable, or a dynlinked library);
- an offset within this block.

When the data is unmarshaled, the runtime looks for a code block with a matching digest and rebuilds a code pointer into this block.  It is assumed that if the digests match, the code pointer thus rebuilt will point to the same function as in the program that marshaled the function value.

Issue #5942 shows that this is not the case.  Currently, the digest for the main executable is just a hash of the code section (the bytecode or machine-code instructions).  However, two functions may have the same code and behave differently because they contain different string literals or other structured constants, which are stored in the data section.  

The first commit in this PR, 8bab0a0, adds a test that reproduces the problem.

@bvaugon proposed fixes in PR #330 and #332 but these PRs stalled in 2015 due to technical issues in native code.  The present PR intends to restart the work and bring it to an adequate fix.

The second commit in this PR, ea2989a, solves the problem for bytecode the way @bvaugon proposed, namely: the digest for the code block of the main executable is the MD5 hash of the bytecode and of the initial global data (the marshaled image of the structured constants stored in the global data table).  I think this addresses the issue in an uncontroversial manner.  I had to refactor a bit the code that loads the bytecode and prepares it for execution, but the result is no worse than before.  In particular, the code and the data sections of the bytecode executable file are loaded in memory exactly once, just like before, but at different points in the runtime system, so that the proper MD5 hash can be computed.

The third and last commit, 261fba8, solves the problem for native code the way I proposed in the discussion of #330 and #332.  It is hopeless to hash the data area of a native-code executable once it's been loaded in memory by the kernel, because, owing to ASLR and other dynamic relocations, its contents can vary from run to run of a program.  

So, let's proceed like we already proceed with dynamically-loaded libraries: use a hash of the whole executable file as the identifier of the main native-code block. A fortiori it includes the code and the initial global data of the executable. It also includes less desirable stuff such as debug info, causing the digest to vary more than absolutely necessary, but this is still safe.
    
The digest could be computed at program start-up, but this might be too costly for a rarely-used feature (marshaling of code pointers).  Instead, we wait until the digest is needed for the first time to read the executable file and compute its MD5 digest.  To avoid problems with relative paths and `chdir()`, the executable file is opened at program start-up and the corresponding file descriptor is kept for later, on-demand computation of the MD5 digest.

The in-memory representation of code fragments was extended and cleaned up so as to 1- include this "lazy computation of digest from file contents" method, 2- factor out the code that computes the digests on demand (function `caml_digest_code_fragment` in runtime/intern.c)

The previous "lazy computation of digest from code area" method remains available and is still used in some cases relative to the toplevel.  It may not be safe to digest only the code and not associated data.  On the other hand it is not obvious what data is associated with that code.  This needs further review.

CI is green, except for ARM in Thumb-2 mode.  This uncovers a problem that has been with us forever but hasn't been noticed yet: ARM Thumb-2 code pointers have the low bit set, hence look like integers to the OCaml runtime, hence are marshaled as integers, i.e. absolute code addresses.  This is wrong in many ways, but will have to be addressed elsewhere, as part of the "no page table" work.  In the meantime I'll look into disabling the test on ARM.

xavierleroy committed:
Test for marshaled closures (issue #5942)

This tests marshaling and unmarshaling of function closures
between two programs that have the same code area but differ
in their data areas.

The test is disabled on ARM because in ARM Thumb2, code pointers have
the low bit set, thus look like integers during marshaling of
closures.  This is wrong and causes the "closures" test to fail.
The issue should go away once we have a "parsable" in-memory format for
function closures.  In the meantime let's just disable this test on ARM.

xavierleroy committed:
Hash both bytecode and global data for the main code fragment

This addresses issue #5942 in the case of bytecode executables.

The checksum associated with the main code fragment covers both
the bytecode and the global data, as read off the bytecode executable.

This way, programs that differ only by their data are considered
incompatible with respect to marshaled function closures.

Also: refactor code between the two bytecode startup functions
caml_main and caml_startup_code_exn.

xavierleroy committed:
Hash the native executable file for the main code fragment

This addresses issue #5942 in the case of native-code compilation.

The checksum associated with the main code fragment is the digest of
the executable file AND of the bytes in the code area.  A fortiori it
includes the code and the initial global data of the executable. It
also includes less desirable stuff such as debug info, causing the
checksum to vary more than absolutely necessary, but this is still
safe.

Besides the executable file, we also include the bytes in the code
area because it could be that the executable file is just a loader for
a shared object that contains the actual OCaml native code.  In this
case, hashing the executable file provides no useful identifier for
the OCaml code, and it would be a regression compared with the old
behavior (of hashing the bytes in the code area).

For this reason, let's hash BOTH the executable file AND the bytes
in the code area.  This avoids the regression in the weird case above,
while giving a hash that accounts for both code and data in the common case.

The checksum could be computed at program start-up, but this might be
too costly for a rarely-used feature (marshaling of code pointers).
Instead, we wait until the checksum is needed for the first time
to read the executable file and compute its MD5 digest.
To avoid problems with relative paths and chdir(), the executable
file is opened at program start-up and the corresponding file descriptor
is kept for later, on-demand computation of the MD5 digest.

The in-memory representation of code fragments was extended and cleaned
up to
1- include this "lazy computation of digest from file contents" method,
2- factor out the code that computes the digests on demand
   (function `caml_digest_code_fragment` in runtime/intern.c)

The previous "lazy computation of digest from code area" method
remains available and is still used in some cases relative to
the toplevel.  It may not be safe to digest only the code and
not associated data.  On the other hand it is not obvious what
data is associated with that code.  This needs further review.

dra27 commented on runtime/startup_nat.c:
```diff
@@ -77,11 +87,39 @@ static void init_static(void)
+     the whole executable file as digest for the main code fragment.
+     This digest is computed on-demand from a file descriptor
+     opened on the executable file. */
+  fd = open_os(exe_name, O_RDONLY | O_BINARY, 0);
```
I haven’t checked the details for how the executable image is locked by Windows while running (it is locked, though), but I’m wondering if the native fd may want opening with the various sharing flags and (possibly) some additional error handling? Alternatively, given that threading is guaranteed present, the hash could be computed in parallel, but this is probably over-thinking it.

xavierleroy replied:
The test passed on several Windows platforms from CI.  (A couple other platforms didn't work, but for unrelated reasons.)  So, I hope the code that reads the executable file seems to work under Windows like it does under Unix.  I'm not sure what you want to achieve with other sharing flags?

dra27 replied:
Indeed - I was referring to the difference on Windows where by default you cannot delete or rename a file which is open (this has stung us elsewhere before - especially when antivirus programs keep files open for a long time after they’ve finished in the testsuite). My memory is that you can rename, but not delete, a running program - it seems worth checking (which I’m happy to do, just not this second!)

dra27 replied:
I can confirm that with this PR, all Windows programs compiled with ocamlopt lose the ability to be renamed while running.

Two options (both of which I'd be happy to contribute) - specify `O_SHARE_DELETE` (which I think requires using `CreateFile` directly and then `_open_osfhandle`) or split this off into something like `caml_init_process_hash` called here which on Unix opens the fd and on Windows spawns a thread to do the hashing straight away and then later `caml_get_process_hash` which on Unix computes the hash and on Windows blocks (if necessary) until the thread finishes. Given that the process will have just been loaded by the OS (and mapped), computing the hash eagerly in parallel at the start should at least not involve (much) disk access, and the ugliness of the difference in behaviour can at least be masked by functions in `unix.c` and `win32.c`.

See [this old comment](https://github.com/ocaml/ocaml/blob/4.07/testsuite/makefiles/Makefile.several#L101-L108) from the pre-ocamltest testsuite for where this has bitten us before.

As an aside - is there enough information at link time in ocamlopt to be able to determine whether marshalling of closures is actually needed in a given program (it's correct I think that if neither Dynlink nor Marshal is actually used, then we know that this functionality is never needed)?

xavierleroy replied:
I'm not sure it is that important to be able to rename or delete executable files while they are running.  But just to avoid a possible regression in this department, I pushed a new version of the PR that opens the executable file in FILE_SHARE_DELETE mode, using `CreateFile` as suggested.

I'm not enthusiastic about computing the hash speculatively in a background thread.  For one thing this would still waste a bit of resources (energy, disk or memory bandwidth) if the program doesn't use marshaling of closures.  For another thing, terminating the thread cleanly when the main program exits is not completely trivial, we ran into this problem in some of the tests for threads.

xavierleroy committed:
Add and use caml_open_file function to open a file in "share delete" mode

Under Windows, the `open` function from the C runtime library opens
the file without the FILE_SHARE_DELETE flag, meaning that the file
cannot be deleted while the file descriptor remains open.

This is inconsistent with the POSIX behavior, and a pain in some
applications, e.g. the "open executable file for later, on-demand hashing"
behavior recently introduced.

This commit adds a `caml_open_file` function to the `<caml/osdeps.h>`
interface that opens the given file path read-only or read-write,
with POSIX / FILE_SHARE_DELETE behavior w.r.t. deletion.

We'll use it soon in the startup code for native executables, in order to
open the executable file.

xavierleroy commented:
The latest version of this PR went through CI successfully. The test that reveals a deeper issue with ARM Thumb2 was made to be skipped on ARM.

alainfrisch commented:
This approach does not seem very robust to me:

  - The executable name might not be the file that contains the OCaml code for the "main OCaml program" (`-output-obj`, linked dynamically into a main program).

  - Is there anything that prevent the content of the executable file to be modified while it is running (I understand the compiler will usually create a new file, not modify existing ones, but are we sure that "cp" or other ways to move file around never modify existing files?).

Why don't we do the work at link-time?  It seems easy enough to include in the data section of images a digest of all OCaml objects/libraries (.o/.a, .obj/.lib) linked into the image.  Are we concerned about the compilation/link-time overhead?  It should be rather small compared to linking itself, and if it takes time, it would also take time to do it at runtime, which seems worse (admittedly, only people relying on marshaling closures would suffer, but they would suffer everytime).

xavierleroy commented:
> The executable name might not be the file that contains the OCaml code for the "main OCaml program" (-output-obj, linked dynamically into a main program).

Yes, you can still shoot yourself in the foot.  (Hence the title of this PR: "Increase safety", not "Ensure safety".)  

> Is there anything that prevent the content of the executable file to be modified while it is running

Yes, Unix kernels will give you a "text file busy" error if you try to modify in-place a file that is executing.  See https://unix.stackexchange.com/questions/138214/how-is-it-possible-to-do-a-live-update-while-a-program-is-running .  I suspect Windows have similar safeguards.

The proper way to update an executable file is to remove it then recreate it.  In this case the old version keeps running.  My PR opens a descriptor on the executable files at the very beginning of execution, so if the remove and recreate happens later, the correct (old) version of the executable file will be hashed.  There is a race condition, though, so, again, you can still shoot yourself in the foot.

> Why don't we do the work at link-time? It seems easy enough to include in the data section of images a digest of all OCaml objects/libraries (.o/.a, .obj/.lib) linked into the image. 

You're welcome to implement this approach and report back.  My gut feeling is that it will be a bigger, more complicated change than what I am proposing here.

I'm trying to address an old issue in an obscure part of the system in the simplest possible way.  We've already talked about fancier solutions in #330 and #332 and nothing came out of it.  Please be a little more supportive.

alainfrisch commented:
I'd like to be supportive, but I want to be convinced that this indeed increases the safety, and that we don't trade a small-enough risk (same code block, different data section) for a bigger one (hashing the wrong file).  Our main application is a small .NET program (which changes rarely), which links dynamically against a DLL that includes the entire OCaml runtime + code (main program).  If we were using closure marshaling (we don't), would we loose the check entirely with the suggested solution?

xavierleroy commented:
> Our main application is a small .NET program (which changes rarely), which links dynamically against a DLL that includes the entire OCaml runtime + code (main program). 

All right, so, it was not a hypothetical what-can-go-wrong scenario :-)  The gun is pointing in the general direction of your feet, indeed.

> I want to be convinced that this indeed increases the safety, and that we don't trade a small-enough risk (same code block, different data section) for a bigger one (hashing the wrong file).

OK.  What about the following: we hash BOTH the contents of the code area (like in the current implementation) AND the contents of the executable file (like in this PR).  It's no worse than the current implementation, even if the wrong file is hashed.  It is better (w.r.t. issue #5942) if the correct file is hashed, which I hope is almost always the case.

alainfrisch commented:
> it was not a hypothetical what-can-go-wrong scenario :-)

It was hypothetical in the sense that we stay away from marshaling closure anyway, but still not a purely rhetorical point just to annoy you :-)

> What about the following: we hash BOTH the contents of the code area (like in the current implementation) AND the contents of the executable file (like in this PR).

This addresses my main concern indeed, thanks!

xavierleroy commented:
We're making progress, that's excellent.  I just pushed a version of the PR that does the double hashing (executable file + code area) mentioned above.  I still need to clean up the test, but reviews are welcome at this point.

xavierleroy committed:
Update Changes

xavierleroy commented:
For the record:

>  I just pushed a version of the PR that does the double hashing (executable file + code area) mentioned above

... and found out that it doesn't work: in native code, the contents of the code area can vary from run to run (owing to ASLR and what not), so the hash is not stable.

6 months later, I'm out of ideas on how to solve this issue, so I'll close this PR.
